src/Path/Internal.hs

```diff
@@ -45,24 +49,38 @@ instance Eq (Path b t) where
 instance Ord (Path b t) where
   compare (Path x) (Path y) = compare x y
 
+curDirNormalizedFP :: FilePath
+curDirNormalizedFP = '.' : [FilePath.pathSeparator]
+
+-- | Convert to a 'FilePath' type.
+--
+-- All directories have a trailing slash, so if you want no trailing
+-- slash, you can use 'System.FilePath.dropTrailingPathSeparator' from
+-- the filepath package.
+toFilePath :: Path b t -> FilePath
+toFilePath (Path []) = curDirNormalizedFP
+toFilePath (Path x)  = x
+
 -- | Same as 'show . Path.toFilePath'.
 --
 -- The following property holds:
 --
 -- @x == y ≡ show x == show y@
 instance Show (Path b t) where
-  show (Path x) = show x
+  show = show . toFilePath
 
 instance NFData (Path b t) where
   rnf (Path x) = rnf x
 
 instance ToJSON (Path b t) where
-  toJSON (Path x) = toJSON x
+  toJSON = toJSON . toFilePath
   {-# INLINE toJSON #-}
 #if MIN_VERSION_aeson(0,10,0)
-  toEncoding (Path x) = toEncoding x
+  toEncoding = toEncoding . toFilePath
   {-# INLINE toEncoding #-}
 #endif
 
 instance Hashable (Path b t) where
-  hashWithSalt n (Path path) = hashWithSalt n path
+  -- hash of "" is just the salt, so use toFilePath to convert
+  -- it to "." instead to get a reasonable hash
+  hashWithSalt n path = hashWithSalt n (toFilePath path)
```

Comment:
It seems you missed the explanation which is already there right above the changed line :-) Is that enough or need to be rephrased for more clarity?